File: python/src/revenant/core/cert_info.py
# pyright: reportUnknownMemberType=false, reportUnknownVariableType=false, reportUnknownArgumentType=false
"""
Certificate information extraction from CMS/PKCS#7 blobs, X.509 certs, and signed PDFs.

Includes identity discovery via enum-certificates (preferred) or dummy-hash
signing (fallback).  Pure data-parsing functions are used by any code that
needs signer info from an existing signature.
"""

from __future__ import annotations

__all__ = [
    "discover_identity_from_server",
    "extract_all_cert_info_from_pdf",
    "extract_cert_info_from_cms",
    "extract_cert_info_from_pdf",
    "extract_cert_info_from_x509",
]

import datetime
import logging
import re
from typing import TYPE_CHECKING

from asn1crypto import cms as asn1_cms
from asn1crypto import x509 as asn1_x509

from ..constants import SHA1_DIGEST_SIZE
from ..errors import AuthError, CertificateError, RevenantError
from .pdf import BYTERANGE_PATTERN, extract_cms_from_byterange_match

if TYPE_CHECKING:
    from ..network.protocol import SigningTransport

_logger = logging.getLogger(__name__)

# OIDs for common subject fields
_OID_CN = "2.5.4.3"
_OID_EMAIL = "1.2.840.113549.1.9.1"
_OID_ORG = "2.5.4.10"


def _extract_info_from_cert_object(cert: asn1_x509.Certificate) -> dict[str, str | None]:
    """Extract CN, email, org, dn from an asn1crypto certificate object.

    Also logs warnings for expired or not-yet-valid certificates.
    """
    subject = cert.subject

    # Warn about certificate validity issues
    try:
        not_before = cert.not_valid_before
        not_after = cert.not_valid_after
        now = datetime.datetime.now(datetime.timezone.utc)
        if not_before and now < not_before:
            _logger.warning("Certificate is not yet valid (notBefore: %s)", not_before)
        elif not_after and now > not_after:
            _logger.warning("Certificate has expired (notAfter: %s)", not_after)
    except (KeyError, TypeError, ValueError) as e:
        _logger.debug("Cannot check certificate validity dates: %s", e)

    # Extract individual fields by OID
    fields: dict[str, str | None] = {"name": None, "email": None, "organization": None}
    oid_map = {_OID_CN: "name", _OID_EMAIL: "email", _OID_ORG: "organization"}

    for rdn in subject.chosen:
        for attr in rdn:
            oid = attr["type"].dotted
            if oid in oid_map:
                fields[oid_map[oid]] = attr["value"].native

    fields["dn"] = subject.human_friendly
    return fields


def extract_cert_info_from_cms(cms_der: bytes) -> dict[str, str | None]:
    """
    Extract signer certificate info from a CMS/PKCS#7 DER blob.

    Uses ``asn1crypto`` for lenient PKCS#7 parsing — handles certificates
    with BMPString-encoded DN fields (e.g. EKENG's CA uses BMPString for CN/O).

    Args:
        cms_der: Raw DER-encoded CMS/PKCS#7 bytes.

    Returns:
        dict with keys: name (CN), email, organization, dn (full subject).

    Raises:
        RevenantError if parsing fails or no certificate found.
    """
    try:
        content_info = asn1_cms.ContentInfo.load(cms_der)
        signed_data = content_info["content"]
        certs = signed_data["certificates"]
    except (ValueError, TypeError, KeyError, OSError) as e:
        raise CertificateError(f"Failed to parse CMS/PKCS#7 blob: {e}") from e

    if not certs:
        raise CertificateError("No certificate subject found in CMS blob.")

    cert = certs[0].chosen
    return _extract_info_from_cert_object(cert)


def extract_cert_info_from_x509(cert_der: bytes) -> dict[str, str | None]:
    """
    Extract signer info from a raw DER-encoded X.509 certificate.

    Args:
        cert_der: Raw DER-encoded X.509 certificate bytes.

    Returns:
        dict with keys: name (CN), email, organization, dn (full subject).

    Raises:
        RevenantError if parsing fails.
    """
    try:
        cert = asn1_x509.Certificate.load(cert_der)
    except (ValueError, TypeError, OSError) as e:
        raise CertificateError(f"Failed to parse X.509 certificate: {e}") from e

    return _extract_info_from_cert_object(cert)


def extract_all_cert_info_from_pdf(pdf_bytes: bytes) -> list[dict[str, str | None]]:
    """
    Extract signer certificate info from ALL signatures in a signed PDF.

    Args:
        pdf_bytes: Raw bytes of a signed PDF.

    Returns:
        list[dict] — one dict per signature, each with keys:
            name (CN), email, organization, dn (full subject).
        Duplicates are removed (same DN).

    Raises:
        RevenantError: If the PDF has no signatures or extraction fails.
    """
    br_matches = list(re.finditer(BYTERANGE_PATTERN, pdf_bytes))
    if not br_matches:
        raise CertificateError("No embedded signature found in this PDF.")

    results = []
    seen_dns: set[str] = set()
    for br in br_matches:
        try:
            cms_der = extract_cms_from_byterange_match(pdf_bytes, br)
            info = extract_cert_info_from_cms(cms_der)
            dn = info.get("dn", "")
            if dn and dn not in seen_dns:
                seen_dns.add(dn)
                results.append(info)
        except RevenantError as exc:  # noqa: PERF203 -- each iteration parses independent CMS blobs; try-except must be per-item
            _logger.debug("Skipping signature (extraction failed): %s", exc)
            continue

    if not results:
        raise CertificateError("Could not extract any certificate info from PDF signatures.")

    return results


def extract_cert_info_from_pdf(pdf_bytes: bytes) -> dict[str, str | None]:
    """
    Extract signer certificate info from a signed PDF.

    If there are multiple signatures, returns the last one.
    Use extract_all_cert_info_from_pdf() to get all signers.

    Args:
        pdf_bytes: Raw bytes of a signed PDF.

    Returns:
        dict with keys: name (CN), email, organization, dn (full subject).

    Raises:
        RevenantError if the PDF has no signature or parsing fails.
    """
    all_info = extract_all_cert_info_from_pdf(pdf_bytes)
    return all_info[-1]


def discover_identity_from_server(
    transport: SigningTransport, username: str, password: str, timeout: int
) -> dict[str, str | None]:
    """
    Discover signer identity from the server.

    Tries enum-certificates first (cleaner, no dummy signing), falls back
    to signing a dummy SHA-1 hash if the server doesn't support it.

    Args:
        transport: Signing transport to use for the SOAP call.
        username: Revenant username.
        password: Revenant password.
        timeout: Request timeout in seconds.

    Returns:
        dict with keys: name (CN), email, organization, dn.

    Raises:
        AuthError: If credentials are wrong.
        RevenantError: If the server rejects the request.
        TLSError: On connection issues.
    """
    # Try enum-certificates (preferred: gets cert directly, no signing)
    url = getattr(transport, "url", None)
    if url:
        try:
            from ..network.soap_transport import enum_certificates

            certs = enum_certificates(url, username, password, timeout)
            if certs:
                _logger.debug("Identity discovered via enum-certificates")
                return extract_cert_info_from_x509(certs[0])
            _logger.debug("enum-certificates returned no certificates")
        except AuthError:
            raise  # Auth errors must propagate immediately
        except (RevenantError, ValueError, TypeError, KeyError, OSError):
            _logger.debug("enum-certificates not available, falling back to dummy-hash")

    # Fallback: sign dummy hash and extract cert from CMS
    _logger.debug("Discovering identity via dummy-hash signing")
    dummy_hash = b"\x00" * SHA1_DIGEST_SIZE
    cms_der = transport.sign_hash(dummy_hash, username, password, timeout)
    return extract_cert_info_from_cms(cms_der)

File: typescript/src/core/cert-info.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Certificate information extraction from CMS/PKCS#7 blobs,
 * X.509 certs, and signed PDFs.
 *
 * Includes identity discovery via enum-certificates (preferred)
 * or dummy-hash signing (fallback).
 */

import * as asn1js from "asn1js";
import * as pkijs from "pkijs";

import { SHA1_DIGEST_SIZE } from "../constants.js";
import { AuthError, CertificateError, RevenantError } from "../errors.js";
import { logger } from "../logger.js";
import type { SigningTransport } from "../network/protocol.js";
import { extractCmsFromByterangeMatch, findByteRanges } from "./pdf/cms-extraction.js";

/** Get a proper ArrayBuffer from a Uint8Array. */
function toArrayBuffer(data: Uint8Array): ArrayBuffer {
  const buf = new ArrayBuffer(data.byteLength);
  new Uint8Array(buf).set(data);
  return buf;
}

// OIDs for common subject fields
const OID_CN = "2.5.4.3";
const OID_EMAIL = "1.2.840.113549.1.9.1";
const OID_ORG = "2.5.4.10";

/**
 * Extract a string value from an ASN.1 value object.
 * Handles Utf8String, PrintableString, BmpString, IA5String, and generic valueBlock.
 */
function extractAsn1StringValue(value: asn1js.BaseBlock): string | null {
  if (
    value instanceof asn1js.Utf8String ||
    value instanceof asn1js.PrintableString ||
    value instanceof asn1js.BmpString ||
    value instanceof asn1js.IA5String
  ) {
    return value.valueBlock.value;
  }
  const block: unknown = value.valueBlock;
  if (block !== null && typeof block === "object" && "value" in block) {
    return String(block.value);
  }
  return null;
}

export interface CertInfo {
  name: string | null;
  email: string | null;
  organization: string | null;
  dn: string | null;
}

/**
 * Extract CN, email, org, dn from a pkijs Certificate object.
 */
function extractInfoFromCertObject(cert: pkijs.Certificate): CertInfo {
  const subject = cert.subject;
  const fields: CertInfo = {
    name: null,
    email: null,
    organization: null,
    dn: null,
  };

  // Check certificate validity
  try {
    const now = new Date();
    const notBefore = cert.notBefore.value;
    const notAfter = cert.notAfter.value;
    if (now < notBefore) {
      logger.warn(`Certificate is not yet valid (notBefore: ${notBefore.toISOString()})`);
    } else if (now > notAfter) {
      logger.warn(`Certificate has expired (notAfter: ${notAfter.toISOString()})`);
    }
  } catch {
    // Cannot check certificate validity dates
  }

  const oidMap: Record<string, keyof CertInfo> = {
    [OID_CN]: "name",
    [OID_EMAIL]: "email",
    [OID_ORG]: "organization",
  };

  // Extract individual fields by OID
  for (const rdn of subject.typesAndValues) {
    const oid = rdn.type;
    const fieldKey = oidMap[oid];
    if (fieldKey) {
      const extracted = extractAsn1StringValue(rdn.value);
      if (extracted) {
        fields[fieldKey] = extracted;
      }
    }
  }

  // Build human-readable DN
  const dnParts: string[] = [];
  for (const rdn of subject.typesAndValues) {
    const strValue = extractAsn1StringValue(rdn.value) ?? "";
    const oidName = getOidName(rdn.type);
    dnParts.push(`${oidName}=${strValue}`);
  }
  fields.dn = dnParts.join(", ");

  return fields;
}

function getOidName(oid: string): string {
  const oidNames: Record<string, string> = {
    "2.5.4.3": "CN",
    "2.5.4.4": "SN",
    "2.5.4.5": "SERIALNUMBER",
    "2.5.4.6": "C",
    "2.5.4.7": "L",
    "2.5.4.8": "ST",
    "2.5.4.10": "O",
    "2.5.4.11": "OU",
    "1.2.840.113549.1.9.1": "E",
  };
  return oidNames[oid] ?? oid;
}

/**
 * Parse a DER-encoded X.509 certificate.
 */
function parseCertificate(certDer: Uint8Array): pkijs.Certificate {
  const asn1 = asn1js.fromBER(toArrayBuffer(certDer));
  if (asn1.offset === -1) {
    throw new CertificateError("Failed to parse ASN.1 structure from certificate");
  }
  return new pkijs.Certificate({ schema: asn1.result });
}

/**
 * Extract signer certificate info from a CMS/PKCS#7 DER blob.
 */
export function extractCertInfoFromCms(cmsDer: Uint8Array): CertInfo {
  let signedData: pkijs.SignedData;
  try {
    const asn1 = asn1js.fromBER(toArrayBuffer(cmsDer));
    if (asn1.offset === -1) {
      throw new Error("Invalid ASN.1");
    }
    const contentInfo = new pkijs.ContentInfo({ schema: asn1.result });
    signedData = new pkijs.SignedData({ schema: contentInfo.content });
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    throw new CertificateError(`Failed to parse CMS/PKCS#7 blob: ${msg}`);
  }

  const certs = signedData.certificates;
  if (!certs || certs.length === 0) {
    throw new CertificateError("No certificate subject found in CMS blob.");
  }

  const cert = certs[0];
  if (cert === undefined || !(cert instanceof pkijs.Certificate)) {
    throw new CertificateError("First certificate in CMS blob is not an X.509 certificate.");
  }

  return extractInfoFromCertObject(cert);
}

/**
 * Extract signer info from a raw DER-encoded X.509 certificate.
 */
export function extractCertInfoFromX509(certDer: Uint8Array): CertInfo {
  try {
    const cert = parseCertificate(certDer);
    return extractInfoFromCertObject(cert);
  } catch (e) {
    if (e instanceof CertificateError) throw e;
    const msg = e instanceof Error ? e.message : String(e);
    throw new CertificateError(`Failed to parse X.509 certificate: ${msg}`);
  }
}

/**
 * Extract signer certificate info from ALL signatures in a signed PDF.
 *
 * Duplicates are removed (same DN).
 */
export function extractAllCertInfoFromPdf(pdfBytes: Uint8Array): CertInfo[] {
  const brMatches = findByteRanges(pdfBytes);
  if (brMatches.length === 0) {
    throw new CertificateError("No embedded signature found in this PDF.");
  }

  const results: CertInfo[] = [];
  const seenDns = new Set<string>();

  for (const br of brMatches) {
    try {
      const cmsDer = extractCmsFromByterangeMatch(pdfBytes, br);
      const info = extractCertInfoFromCms(cmsDer);
      const dn = info.dn ?? "";
      if (dn && !seenDns.has(dn)) {
        seenDns.add(dn);
        results.push(info);
      }
    } catch (exc) {
      if (exc instanceof RevenantError) continue;
      throw exc;
    }
  }

  if (results.length === 0) {
    throw new CertificateError("Could not extract any certificate info from PDF signatures.");
  }

  return results;
}

/**
 * Extract signer certificate info from a signed PDF.
 *
 * If there are multiple signatures, returns the last one.
 */
export function extractCertInfoFromPdf(pdfBytes: Uint8Array): CertInfo {
  const allInfo = extractAllCertInfoFromPdf(pdfBytes);
  const lastInfo = allInfo[allInfo.length - 1];
  if (lastInfo === undefined) {
    throw new CertificateError("Could not extract any certificate info from PDF signatures.");
  }
  return lastInfo;
}

/**
 * Discover signer identity from the server.
 *
 * Tries enum-certificates first (cleaner, no dummy signing), falls back
 * to signing a dummy SHA-1 hash if the server doesn't support it.
 */
export async function discoverIdentityFromServer(
  transport: SigningTransport,
  username: string,
  password: string,
  timeout: number,
): Promise<CertInfo> {
  // Try enum-certificates (preferred)
  const url = transport.url;
  if (url) {
    try {
      const { enumCertificates } = await import("../network/soap-transport.js");
      const certs = await enumCertificates(url, username, password, timeout);
      const firstCert = certs[0];
      if (firstCert) {
        return extractCertInfoFromX509(firstCert);
      }
    } catch (exc) {
      if (exc instanceof AuthError) throw exc;
      // enum-certificates not available, fall back to dummy-hash
    }
  }

  // Fallback: sign dummy hash and extract cert from CMS
  const dummyHash = new Uint8Array(SHA1_DIGEST_SIZE);
  const cmsDer = await transport.signHash(dummyHash, username, password, timeout);
  return extractCertInfoFromCms(cmsDer);
}
